runtime-common/loader.ts

```diff
@@ -0,0 +1,212 @@
+// @ts-ignore
+import TransformModulesAmd from "@babel/plugin-transform-modules-amd";
+import { transformSync } from "@babel/core";
+import { Deferred } from "./deferred";
+import type { Realm } from "@cardstack/runtime-common/realm";
+import { RealmPaths } from "./paths";
+import { baseRealm } from "@cardstack/runtime-common";
+
+type RegisteredModule = {
+  state: "registered";
+  dependencyList: string[];
+  implementation: Function;
+};
+
+type FetchingModule = {
+  state: "fetching";
+  deferred: Deferred<unknown>;
+};
+
+type Module =
+  | RegisteredModule
+  | FetchingModule
+  | {
+      state: "preparing";
+      implementation: Function;
+      moduleInstancePromise: Promise<object>;
+    }
+  | {
+      state: "evaluated";
+      moduleInstance: object;
+    }
+  | {
+      state: "broken";
+      exception: any;
+    };
+
+export class Loader {
+  private modules = new Map<string, Module>();
+  private realmPath: RealmPaths;
+
+  constructor(
+    private realm: Realm,
+    private openFile: (url: URL) => Promise<string>
+  ) {
+    this.realmPath = new RealmPaths(realm.url);
+  }
+
+  async load<T extends object>(moduleIdentifier: string): Promise<T> {
+    if (!moduleIdentifier.startsWith("http")) {
+      throw new Error(
+        `expected module identifier to be a URL: "${moduleIdentifier}"`
+      );
+    }
+    moduleIdentifier = this.resolveModule(moduleIdentifier);
+
+    let module = this.modules.get(moduleIdentifier);
+    if (!module) {
+      module = {
+        state: "fetching",
+        deferred: new Deferred<T>(),
+      };
+      this.modules.set(moduleIdentifier, module);
+      return await this.fetchModule(moduleIdentifier, module);
+    }
+
+    switch (module.state) {
+      case "fetching":
+        return await (module.deferred as Deferred<T>).promise;
+      case "preparing":
+        return (await module.moduleInstancePromise) as T;
```

Comment:
this was a critical adjustment that I made to the state machine from the original source in cardhost. Previously, in cardhost, this used to fall though to the `evaluated` state and just return the `module.moduleInstance`. After adding a new state that comes before this one ("fetching"), i was seeing that the dependency loading would short circuit if the module was in a preparing state and return the unbuilt `moduleInstance` (still an empty object). I changed the state variable for the "preparing" state to be a promise for the `moduleInstance` which I then resolved after the dependencies for the module were all loaded. However, I think this probably breaks the cyclic module loading...